crates/circuit/src/dag_circuit.rs

```diff
@@ -2551,25 +2553,48 @@ def _format(operand):
     fn swap_nodes(&mut self, node1: &DAGNode, node2: &DAGNode) -> PyResult<()> {
-        // if not (isinstance(node1, DAGOpNode) and isinstance(node2, DAGOpNode)):
-        //     raise DAGCircuitError("nodes to swap are not both DAGOpNodes")
-        // try:
-        //     connected_edges = self._multi_graph.get_all_edge_data(node1._node_id, node2._node_id)
-        // except rx.NoEdgeBetweenNodes as no_common_edge:
-        //     raise DAGCircuitError("attempt to swap unconnected nodes") from no_common_edge
-        // node1_id = node1._node_id
-        // node2_id = node2._node_id
-        // for edge in connected_edges[::-1]:
-        //     edge_find = lambda x, y=edge: x == y
-        //     edge_parent = self._multi_graph.find_predecessors_by_edge(node1_id, edge_find)[0]
-        //     self._multi_graph.remove_edge(edge_parent._node_id, node1_id)
-        //     self._multi_graph.add_edge(edge_parent._node_id, node2_id, edge)
-        //     edge_child = self._multi_graph.find_successors_by_edge(node2_id, edge_find)[0]
-        //     self._multi_graph.remove_edge(node1_id, node2_id)
-        //     self._multi_graph.add_edge(node2_id, node1_id, edge)
-        //     self._multi_graph.remove_edge(node2_id, edge_child._node_id)
-        //     self._multi_graph.add_edge(node1_id, edge_child._node_id, edge)
-        todo!()
+        let node1 = node1.node.unwrap();
+        let node2 = node2.node.unwrap();
+        let dag_binding = self.dag.clone();
+
+        // extracted functionality from rustworkx's 'get_edge_data'
+        let connected_edges = dag_binding
+            .edges(node1)
+            .filter(|edge| edge.target() == node2)
+            .collect::<Vec<_>>();
+
+        // find first parent/child for a given edge, depending on the specified direction:
+        //  - Incoming -> parent
+        //  - Outgoing -> child
+        // merged functionality from rustworkx's 'find_predecessors_by_edge' and 'find_successors_by_edge'
+        // to understand name, see: https://en.wikipedia.org/wiki/Oyakodon
+        let find_first_oyako_by_edge =
+            |(node, direction, ref_edge): (NodeIndex, Direction, EdgeReference<Wire>)| {
+                let mut oyakosan = Vec::new();
+                let raw_edges = dag_binding.edges_directed(node, direction);
+                for edge in raw_edges {
+                    if ref_edge.weight() == edge.weight() {
+                        oyakosan.push(edge);
+                        break;
+                    }
+                }
+                oyakosan[0]
+            };
+
+        for edge in connected_edges.iter().rev() {
+            let edge_parent = find_first_oyako_by_edge((node1, Incoming, *edge));
+            self.dag.remove_edge(edge_parent.id());
+            self.dag
+                .add_edge(edge_parent.source(), node2, *edge.weight());
+            self.dag
+                .remove_edge(self.dag.edges_connecting(node1, node2).collect::<Vec<_>>()[0].id());
```

Comment:
This line looks problematic. We ought to specifically remove the current wire between `node1` and `node2`. 

Technically, what you have might work since we'll have removed all wires connecting these two nodes by the time the loop finishes, but it's quite cryptic since we are iterating through specific wires, but removing an arbitrary wire in each iteration.

Instead, maybe you can call the helper function:

```rust
self.dag.remove_edge(... find_first_oyako_by_edge((node1, Outgoing, *edge)) ...)
```